### DISCUSSION TITLE
Template covariance / `Collection<int, Model>` / `Collection<int, *>` / WHAT??

### DISCUSSION THREAD

**Discussion by rudiedirkx** [larastan/larastan#2093] [Q&A]
I understand covariance and contravariance in PHP classes/methods/functions. I apparently do not understand it in generics.

## `#1`

I have a class `OrderUpdater` that works for any model. It takes a `ModelCollection` of any model, and updates the models within. Constructor bit:

```php
class OrderUpdater {

	/** @var ModelCollection<int, Model> */
	protected ModelCollection $models;

	/** @param ModelCollection<int, Model> $models */
	public function __construct(ModelCollection $models) {
		$this->models = $models->keyBy('id');
	}

}
```

(The smart people probably immediately see the problem with `ModelCollection<int, Model>` because it never takes literal `Model` objects.)

My controller makes an OrderUpdater:

```php
$tags = $organization->tags;
\PHPStan\dumpType($tags); // App\Base\ModelCollection<int, App\Models\Tag>

(new OrderUpdater($tags))->updateWithTransaction((array) $request['o']);
```

And then PhpStan complains:

> `Parameter #1 $models of class App\Base\OrderUpdater constructor expects App\Base\ModelCollection<int, App\Base\Model>, App\Base\ModelCollection<int, App\Models\Tag> given.`
> - 💡 Template type TModel on class App\Base\ModelCollection is not covariant. Learn more:
> - https://phpstan.org/blog/whats-up-with-template-covariant

I've read that page 6 times now, and I definitely don't get it. Why doesn't it accept a `Tag` if it needs a `Model`?? (That's probably not what it means, but that's what it looks like to me.)

## `#2`

So I thought, screw it, I'll just make `ModelCollection`'s `TModel` covariant:

```
/**
 * @template TKey of array-key
 * @template-covariant TModel of Model
 *
 * @extends EloquentCollection<TKey, TModel>
 */
class ModelCollection extends EloquentCollection {
```

but now PhpStan complains about that covariant template!

> `Template type TModel is declared as covariant, but occurs in invariant position in extended type Illuminate\Database\Eloquent\Collection<TKey of (int|string), TModel of App\Base\Model> of class App\Base\ModelCollection.`

So that's also not how that works.

## `#3`

I found out `ModelCollection<int, *>` is a thing earlier, so that's also an option I thought. Instead of `OrderUpdater` requiring `Model` it could require `*`:

```php
class OrderUpdater {

	/** @var ModelCollection<int, *> */
	protected ModelCollection $models;

	/** @param ModelCollection<int, *> $models */
	public function __construct(ModelCollection $models) {
		$this->models = $models->keyBy('id');
	}

}
```

and then everything works for **creating** the `OrderUpdater` object, BUT not when the class actually acts on the object:

```php
$this->models[$id]->update(['o' => $o * 10]);
```

because apparently `*` doesn't mean 'any `Model`':

> `Cannot call method update() on mixed.`

but `mixed`?? Really?? Then what is the point of `*`? Of course it's a `Model`, that's the requirement in the `@template`!

----

I'm losing my mind. Please save me, explain to me in little baby language what generic covariance is, and how I can make my ModelCollection AND `OrderUpdater` work.

**Comment by rudiedirkx** [comment]
Sometimes I even get this beauty of an error (when `@template` **not** covariant), and then I just want to give up because I'm SO dumb I think this is a nonsense error:

> `Method App\Base\ModelCollection::relation() should return App\Base\ModelCollection<int, App\Base\Model> but returns App\Base\ModelCollection<int, App\Base\Model>.`

Are you kidding me??? What is going on inside PhpStan??

**Reply by calebdw** [reply]
This looks like a PHPStan bug, try to replicate it on https://phpstan.org/try and open an issue over there

**Reply by rudiedirkx** [reply]
I don't even know if it's a bug, it looks silly to me, but maybe it makes total sense, because I don't understand generics 😆 I fixed `relation()` so I don't even know how that happened anymore.

**Comment by rudiedirkx** [comment]
I fixed `relation` (no more errors at level 6 anyway):

```php
/** @return self<int, Model> */
public function relation(string $relationName) : self {
	$loaded = $this->pluck($relationName);

	if ($loaded->first() instanceof self) {
		$loaded = $loaded->flatten(1);
	}

	if ($loaded->first() instanceof Model) {
		return $loaded->first()->newCollection($loaded->all());
	}

	return new self();
}
```

A custom PhpStan extension takes care of the real return type, eg. `ModelCollection<int, Tag>`, from the `$relationName`.

**Comment by rudiedirkx** [comment]
Maybe I do understand https://phpstan.org/blog/whats-up-with-template-covariant That's actually pretty well explained. So I want to promise PhpStan that I won't mutate the collection, it's just for 'getting'. The article says:

    /** @param Collection<covariant Animal> $animals */

and even though **that** doesn't seem logical to me, the explanation does, AND it works in my code. 👍
